src/sources/journald.rs

```diff
@@ -303,20 +317,20 @@ impl JournalSource for Journalctl {
     }
 }
 
-impl Iterator for Journalctl {
-    type Item = Result<String, io::Error>;
-    fn next(&mut self) -> Option<Self::Item> {
-        let mut line = Vec::<u8>::new();
-        match self.stdout.read_until(b'\n', &mut line) {
-            Ok(0) => None,
-            Ok(_) => Some(Ok(String::from_utf8_lossy(&line).into())),
-            Err(err) => Some(Err(err)),
-        }
+impl Stream for Journalctl {
+    type Item = io::Result<String>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        Poll::Ready(match ready!(self.project().inner.poll_next(cx)) {
+            Some(Ok(segment)) => Some(Ok(String::from_utf8_lossy(&segment).into())),
```

Comment:
This `.into()` add an unnecessary clone every time `segment` is valid utf8, right? I'm not sure if rust is smart enough to optimize it. 

We can get away with something like
```rust
fn from_utf8_lossy(bytes: Vec<u8>) -> String {
    use std::borrow::Cow;
    match String::from_utf8_lossy(&bytes) {
        Cow::Owned(s) => s,
        _ => unsafe { String::from_utf8_unchecked(bytes) },
    }
}
```
but it add an unsafe block.

Reply:
Yeah, that's looks correct @juchiast, `.into()` will cause extra copy for each valid utf8. I do not think that this copy super critical but it would be nice to change if not unsafe block :disappointed:

Reply:
It is possible without a clone, but does the UTF-8 validation twice (assumes `segment: Vec<u8>`):
```rust
String::from_utf8(segment).unwrap_or_else(|error| String::from_utf8_lossy(error.as_bytes()).into())
```
This is *probably* more efficient, as the majority of the time the source will be valid UTF-8, so it avoids the allocation and copy, and if it isn't valid, it's likely to fail early anyways.